### PR TITLE
[diem-types] Change module publishing txn to take a bundle

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

This will be a breaking change!

Change module publishing transaction from taking one single module to a list of modules the sender want to publish. This will significantly improve the usability of friend feature as friend modules can now be bundled together and publish in one transaction instead of having to separate them into three transactions.

Also breaking change is fine in this case because we never use module publishing transaction before in the system. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

All the existing tests for publishing single transaction works. Will add test in `e2e-testsuite` for publishing multiple transactions.

### CODE REVIEW

**Comment by xli** [comment]
LGTM for the API changes.

**Review by tzakian — APPROVED** [review]
LGTM!

**Comment by runtian-zhou** [comment]
Justification for the breaking change:
1. We never had such transaction in mainnet because module publishing transaction can only be sent by diem root with our current setup. And diem root never implemented such capability. This is why even though it's a breaking change it would be fine for the system.
2. We would be able to get rid of the old module publishing code that was never used by just changing the variant definition.

**Comment by runtian-zhou** [comment]
cc @bmwill should I change the version number for `diem-types` here?

**Comment by runtian-zhou** [comment]
Based on the discussion I don't think we need to update the version number here. That can go to the release of next sdk.

File: api/src/tests/transactions_test.rs [hunk 1]
```diff
@@ -888,7 +892,9 @@ async fn test_signing_message_with_module_payload() {
     );
     let payload = json!({
             "type": "module_payload",
```
Comment (xli) [inline api/src/tests/transactions_test.rs:890]: Same here should match new name

**Review by xli — CHANGES_REQUESTED** [review]

**Comment by bmwill** [comment]
> cc @bmwill should I change the version number for `diem-types` here?

Nope, no need to do that now

File: api/doc/openapi.yaml [hunk 2]
```diff
@@ -1161,6 +1161,13 @@ components:
...
           items:
             $ref: '#/components/schemas/MoveValue'
+    ModuleBundlePayload:
+      type: object
+      properties:
+        modules:
+          type: array
+          items:
+            $ref: '#/components/schemas/MoveModuleBytecodePayload'
     MoveModuleBytecodePayload:
```
Comment (xli) [inline api/doc/openapi.yaml:1171]: for this `MoveModuleBytecodePayload`, you need remove `type` field. and add a type field to `ModuleBundlePayload` and set example: module_bundle_payload

Comment (xli) [reply api/doc/openapi.yaml:1171]: also `MoveModuleBytecodePayload` is not a payload anymore, probably rename it to `MoveModuleBytecode`

Comment (xli) [reply api/doc/openapi.yaml:1171]: actually, I found we have `MoveModuleBytecode`, so you should just remove `MoveModuleBytecodePayload` after replaced by `ModuleBundlePayload`

File: api/doc/openapi.yaml [hunk 3]
```diff
@@ -1161,19 +1161,13 @@ components:
...
-    MoveModuleBytecodePayload:
+    ModuleBundlePayload:
       type: object
-      required:
-        - type
-        - bytecode
       properties:
-        type:
-          type: string
-          example: move_module_bytecode_payload
```
Comment (xli) [inline api/doc/openapi.yaml:1172]: need type
example: module_bundle_payload

File: api/doc/openapi.yaml [hunk 4]
```diff
@@ -1161,19 +1161,13 @@ components:
           type: array
           items:
             $ref: '#/components/schemas/MoveValue'
-    MoveModuleBytecodePayload:
+    ModuleBundlePayload:
       type: object
-      required:
-        - type
-        - bytecode
```
Comment (xli) [inline api/doc/openapi.yaml:1168]:
```
required:
  - type
  - modules
```

Comment (runtian-zhou) [reply api/doc/openapi.yaml:1168]: Fixed!

File: api/doc/openapi.yaml (hunk 3 above)
Comment (runtian-zhou) [reply api/doc/openapi.yaml:1172]: Fixed!

**Review by xli — DISMISSED** [review]

**Pull request closed without merging**